spec/services/interaction_tracking_service_spec.rb

```diff
@@ -11,4 +11,143 @@
+    let!(:tax_return_2) { create(:tax_return, assigned_user_id: user_no_notifications.id, year: (Rails.configuration.product_year - 1), client: client) }
+    let(:user) { create(:admin_user, email_notification: "yes") }
+    let(:user_no_notifications) { create(:admin_user, email_notification: "no") }
+
+    before do
+      allow(Flipper).to receive(:enabled?).with(:hub_email_notifications).and_return(true)
+      allow(InternalEmail).to receive(:create!).and_call_original
+      allow(SendInternalEmailJob).to receive(:perform_later)
+    end
+
+    it "sends a notification, enqueues the email job and creates an Internal Email" do
+      described_class.record_incoming_interaction(client, set_flag: true, message_received_at: fake_time)
+      expect(InternalEmail).to have_received(:create!).with(
+        mail_class: UserMailer,
+        mail_method: :incoming_interaction_notification_email,
+        mail_args: ActiveJob::Arguments.serialize(
+          client: client,
+          user: user,
+          message_received_at: fake_time
+        )
+      )
+      expect(SendInternalEmailJob).to have_received(:perform_later)
+      expect(InternalEmail.last.mail_class).to eq "UserMailer"
+      expect(InternalEmail.last.mail_method).to eq "incoming_interaction_notification_email"
+    end
+
+    it "doesn't send a message for the user that has chosen to opt-out of email notifications" do
+      described_class.record_incoming_interaction(client, set_flag: true, message_received_at: fake_time)
+      expect(InternalEmail).not_to have_received(:create!).with(
+        mail_class: UserMailer,
+        mail_method: :incoming_interaction_notification_email,
+        mail_args: ActiveJob::Arguments.serialize(
+          client: client,
+          user: user_no_notifications,
+          message_received_at: fake_time
+        )
+      )
+    end
+
+    context "when the interaction type is not client_message" do
+      it "doesn't send any email notifications" do
+        described_class.record_incoming_interaction(client, set_flag: true, message_received_at: nil)
+        expect(InternalEmail).not_to have_received(:create!)
+        expect(SendInternalEmailJob).not_to have_received(:perform_later)
+      end
+    end
+
+    context "when the client doesn't have any assigned users" do
+      before do
+        tax_return_1.update!(assigned_user: nil)
+        tax_return_2.update!(assigned_user: nil)
+      end
+      it "doesn't send any email notifications" do
+        described_class.record_incoming_interaction(client, set_flag: true, message_received_at: fake_time)
+        expect(InternalEmail).not_to have_received(:create!)
+        expect(SendInternalEmailJob).not_to have_received(:perform_later)
+      end
+    end
+
+    context "when only one tax return has an assigned user" do
+      before do
+        tax_return_2.update!(assigned_user: nil)
+      end
```

Comment:
can you explain what this case is testing?
the tax return 2 no longer has an assigned user, but then we check if we have created InternalEmail for the other tax return?

Reply:
Ooo yeah you're right it's not clear, I guees I could add an expect(InternalEmail).not_to have_received(:create!) with the user_no_notifications argument? I guess I wanted to test this line `users_to_contact` but not sure about this best way to do that